crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -805,25 +817,81 @@ macro_rules! impl_system_function {
         #[allow(non_snake_case)]
         impl<Input, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(In<Input>, $($param,)*) -> Out> for Func
         where
-        for <'a> &'a mut Func:
+            for <'a> &'a mut Func:
                 FnMut(In<Input>, $($param),*) -> Out +
-                FnMut(In<Input>, $(SystemParamItem<$param>),*) -> Out, Out: 'static
+                FnMut(In<Input>, $(SystemParamItem<$param>),*) -> Out,
+            Input: 'static,
```

Comment:
Was there no way to make this generic over any `SystemInput`? It would be nice to avoid having to duplicate these impls for each `SystemInput`.

Reply:
Unfortunately not: we would need negative trait implementations to guarantee that `SystemInput` and `SystemParam` are mutually exclusive. Because we can't do that, the trait bounds are theoretically the same for the impl with `SystemInput` and the impl without a `SystemInput`.

Reply:
What if the marker type on `SystemParamFunction` was changed to be different based on whether the first argument is an input or a parameter?

Reply:
I think I have a prototype for this, though unfortunately it does break inference in a doc test which I had to fix by introducing `StaticIn`, the equivalent of `StaticSystemParam` but for `SystemInput`. This might not matter in practice though.

<details>

```diff
diff --git a/crates/bevy_ecs/src/system/function_system.rs b/crates/bevy_ecs/src/system/function_system.rs
index 1a17adbe9d..a8fe973e2c 100644
--- a/crates/bevy_ecs/src/system/function_system.rs
+++ b/crates/bevy_ecs/src/system/function_system.rs
@@ -5,8 +5,8 @@ use crate::{
     query::{Access, FilteredAccessSet},
     schedule::{InternedSystemSet, SystemSet},
     system::{
-        check_system_change_tick, input::SystemInput, InMut, InRef, ReadOnlySystemParam, System,
-        SystemIn, SystemParam, SystemParamItem,
+        check_system_change_tick, input::SystemInput, ReadOnlySystemParam, System, SystemIn,
+        SystemParam, SystemParamItem,
     },
     world::{unsafe_world_cell::UnsafeWorldCell, DeferredWorld, World, WorldId},
 };
@@ -720,12 +720,13 @@ where
 /// use std::num::ParseIntError;
 ///
 /// use bevy_ecs::prelude::*;
+/// use bevy_ecs::system::StaticIn;
 ///
 /// /// Pipe creates a new system which calls `a`, then calls `b` with the output of `a`
 /// pub fn pipe<A, B, AMarker, BMarker>(
 ///     mut a: A,
 ///     mut b: B,
-/// ) -> impl FnMut(<A::In as SystemInput>::Param<'_>, ParamSet<(A::Param, B::Param)>) -> B::Out
+/// ) -> impl FnMut(StaticIn<A::In>, ParamSet<(A::Param, B::Param)>) -> B::Out
 /// where
 ///     // We need A and B to be systems, add those bounds
 ///     A: SystemParamFunction<AMarker>,
@@ -733,8 +734,8 @@ where
 ///     for<'a> B::In: SystemInput<Inner<'a> = A::Out>,
 /// {
 ///     // The type of `params` is inferred based on the return of this function above
-///     move |a_in, mut params| {
-///         let shared = a.run(<A::In as SystemInput>::into_inner(a_in), params.p0());
+///     move |StaticIn(a_in), mut params| {
+///         let shared = a.run(a_in, params.p0());
 ///         b.run(shared, params.p1())
 ///     }
 /// }
@@ -784,6 +785,10 @@ pub trait SystemParamFunction<Marker>: Send + Sync + 'static {
     ) -> Self::Out;
 }
 
+/// A marker type used to distinguish function systems with and without input.
+#[doc(hidden)]
+pub struct HasSystemInput;
+
 macro_rules! impl_system_function {
     ($($param: ident),*) => {
         #[allow(non_snake_case)]
@@ -815,86 +820,32 @@ macro_rules! impl_system_function {
         }
 
         #[allow(non_snake_case)]
-        impl<Input, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(In<Input>, $($param,)*) -> Out> for Func
-        where
-            for <'a> &'a mut Func:
-                FnMut(In<Input>, $($param),*) -> Out +
-                FnMut(In<Input>, $(SystemParamItem<$param>),*) -> Out,
-            Input: 'static,
-            Out: 'static
-        {
-            type In = In<Input>;
-            type Out = Out;
-            type Param = ($($param,)*);
-            #[inline]
-            fn run(&mut self, input: Input, param_value: SystemParamItem< ($($param,)*)>) -> Out {
-                #[allow(clippy::too_many_arguments)]
-                fn call_inner<Input, Out, $($param,)*>(
-                    mut f: impl FnMut(In<Input>, $($param,)*)->Out,
-                    input: Input,
-                    $($param: $param,)*
-                )->Out{
-                    f(In(input), $($param,)*)
-                }
-                let ($($param,)*) = param_value;
-                call_inner(self, input, $($param),*)
-            }
-        }
-
-        #[allow(non_snake_case)]
-        impl<Input, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(InRef<'static, Input>, $($param,)*) -> Out> for Func
+        impl<Input, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<(HasSystemInput, fn(Input, $($param,)*) -> Out)> for Func
         where
             for <'a> &'a mut Func:
-                FnMut(InRef<'_, Input>, $($param),*) -> Out +
-                FnMut(InRef<'_, Input>, $(SystemParamItem<$param>),*) -> Out,
-            Input: ?Sized + 'static,
+                FnMut(Input, $($param),*) -> Out +
+                FnMut(<Input as SystemInput>::Param<'_>, $(SystemParamItem<$param>),*) -> Out,
+            Input: SystemInput + 'static,
             Out: 'static
         {
-            type In = InRef<'static, Input>;
+            type In = Input;
             type Out = Out;
             type Param = ($($param,)*);
             #[inline]
-            fn run(&mut self, input: &Input, param_value: SystemParamItem< ($($param,)*)>) -> Out {
+            fn run(&mut self, input: <Input as SystemInput>::Inner<'_>, param_value: SystemParamItem< ($($param,)*)>) -> Out {
                 #[allow(clippy::too_many_arguments)]
-                fn call_inner<Input: ?Sized, Out, $($param,)*>(
-                    mut f: impl FnMut(InRef<'_, Input>, $($param,)*)->Out,
-                    input: &Input,
+                fn call_inner<I, Out, $($param,)*>(
+                    mut f: impl FnMut(I, $($param,)*)->Out,
+                    input: I,
                     $($param: $param,)*
                 )->Out{
-                    f(InRef(input), $($param,)*)
+                    f(input, $($param,)*)
                 }
                 let ($($param,)*) = param_value;
-                call_inner(self, input, $($param),*)
+                call_inner(self, Input::into_param(input), $($param),*)
             }
         }
-
-        #[allow(non_snake_case)]
-        impl<Input, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(InMut<'static, Input>, $($param,)*) -> Out> for Func
-        where
-            for <'a> &'a mut Func:
-                FnMut(InMut<'_, Input>, $($param),*) -> Out +
-                FnMut(InMut<'_, Input>, $(SystemParamItem<$param>),*) -> Out,
-            Input: ?Sized + 'static,
-            Out: 'static
-        {
-            type In = InMut<'static, Input>;
-            type Out = Out;
-            type Param = ($($param,)*);
-            #[inline]
-            fn run(&mut self, input: &mut Input, param_value: SystemParamItem< ($($param,)*)>) -> Out {
-                #[allow(clippy::too_many_arguments)]
-                fn call_inner<Input: ?Sized, Out, $($param,)*>(
-                    mut f: impl FnMut(InMut<'_, Input>, $($param,)*)->Out,
-                    input: &mut Input,
-                    $($param: $param,)*
-                )->Out{
-                    f(InMut(input), $($param,)*)
-                }
-                let ($($param,)*) = param_value;
-                call_inner(self, input, $($param),*)
-            }
-        }
-    };
+    }
 }
 
 // Note that we rely on the highest impl to be <= the highest order of the tuple impls
diff --git a/crates/bevy_ecs/src/system/input.rs b/crates/bevy_ecs/src/system/input.rs
index 32cac0af18..42962b6c0c 100644
--- a/crates/bevy_ecs/src/system/input.rs
+++ b/crates/bevy_ecs/src/system/input.rs
@@ -180,3 +180,19 @@ impl<E: 'static, B: Bundle> SystemInput for Trigger<'_, E, B> {
         this
     }
 }
+
+/// TODO
+pub struct StaticIn<'a, T: SystemInput>(pub T::Inner<'a>);
+
+impl<'a, T: SystemInput> SystemInput for StaticIn<'a, T> {
+    type Param<'i> = StaticIn<'i, T>;
+    type Inner<'i> = T::Inner<'i>;
+
+    fn into_inner(this: Self::Param<'_>) -> Self::Inner<'_> {
+        this.0
+    }
+
+    fn into_param(this: Self::Inner<'_>) -> Self::Param<'_> {
+        StaticIn(this)
+    }
+}
diff --git a/crates/bevy_ecs/src/system/observer_system.rs b/crates/bevy_ecs/src/system/observer_system.rs
index fac3f079a1..7cae84d160 100644
--- a/crates/bevy_ecs/src/system/observer_system.rs
+++ b/crates/bevy_ecs/src/system/observer_system.rs
@@ -1,8 +1,6 @@
-use bevy_utils::all_tuples;
-
 use crate::{
     prelude::{Bundle, Trigger},
-    system::{System, SystemParam, SystemParamFunction, SystemParamItem},
+    system::System,
 };
 
 use super::IntoSystem;
@@ -55,37 +53,6 @@ where
     }
 }
 
-macro_rules! impl_system_function {
-    ($($param: ident),*) => {
-        #[allow(non_snake_case)]
-        impl<E: 'static, B: Bundle, Out, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(Trigger<E, B>, $($param,)*)> for Func
-        where
-        for <'a> &'a mut Func:
-                FnMut(Trigger<E, B>, $($param),*) -> Out +
-                FnMut(Trigger<E, B>, $(SystemParamItem<$param>),*) -> Out, Out: 'static
-        {
-            type In = Trigger<'static, E, B>;
-            type Out = Out;
-            type Param = ($($param,)*);
-            #[inline]
-            fn run(&mut self, input: Trigger<E, B>, param_value: SystemParamItem< ($($param,)*)>) -> Out {
-                #[allow(clippy::too_many_arguments)]
-                fn call_inner<E: 'static, B: Bundle, Out, $($param,)*>(
-                    mut f: impl FnMut(Trigger<E, B>, $($param,)*) -> Out,
-                    input: Trigger<E, B>,
-                    $($param: $param,)*
-                ) -> Out{
-                    f(input, $($param,)*)
-                }
-                let ($($param,)*) = param_value;
-                call_inner(self, input, $($param),*)
-            }
-        }
-    }
-}
-
-all_tuples!(impl_system_function, 0, 16, F);
-
 #[cfg(test)]
 mod tests {
     use crate::{
```
</details>

Reply:
Fantastic! Thanks for the investigation. I think it's fine for the uncommon case where someone will need `StaticSystemInput` just like they would `StaticSystemParam`.